File: app/src/main/java/com/arnaudpiroelle/manga/ui/history/HistoryFragment.java
package com.arnaudpiroelle.manga.ui.history;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ListView;

import com.arnaudpiroelle.manga.R;
import com.arnaudpiroelle.manga.core.adapter.BaseAdapter;
import com.arnaudpiroelle.manga.core.ui.presenter.Presenter;
import com.arnaudpiroelle.manga.model.History;

import java.util.List;

import javax.inject.Inject;

import butterknife.Bind;
import butterknife.ButterKnife;
import de.greenrobot.event.EventBus;

import static com.arnaudpiroelle.manga.MangaApplication.GRAPH;

public class HistoryFragment extends Fragment implements HistoryPresenter.HistoryListingCallback, SwipeRefreshLayout.OnRefreshListener {

    @Inject EventBus eventBus;

    @Bind(R.id.swipe_refresh) SwipeRefreshLayout swipeRefreshLayout;
    @Bind(R.id.list_history) ListView listView;
    @Bind(R.id.history_empty) View emptyView;

    HistoryPresenter presenter;
    BaseAdapter<History, HistoryView> adapter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        GRAPH.inject(this);

        setHasOptionsMenu(true);

        presenter = new HistoryPresenter(this);
        adapter = new BaseAdapter<>(getActivity(), R.layout.item_view_history);
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_listing_history, container, false);

        ButterKnife.bind(this, view);

        listView.setEmptyView(emptyView);
        listView.setAdapter(adapter);
        swipeRefreshLayout.setOnRefreshListener(this);

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();

        getActivity().setTitle(R.string.title_history);

        presenter.list();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);

        inflater.inflate(R.menu.menu_history, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_clean_history:
                cleanHistory();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void cleanHistory() {
        presenter.cleanHistory();
    }

    @Override
    public void onListingLoading() {
        swipeRefreshLayout.setRefreshing(true);
    }

    @Override
    public void onListingLoaded(List<History> histories) {
        swipeRefreshLayout.setRefreshing(false);

        adapter.setData(histories);
    }

    @Override
    public void onRefresh() {
        presenter.list();
    }
}

File: app/src/main/java/com/arnaudpiroelle/manga/ui/history/HistoryFragment.kt
package com.arnaudpiroelle.manga.ui.history

import android.os.Bundle
import android.view.*
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import com.arnaudpiroelle.manga.R
import com.arnaudpiroelle.manga.core.utils.bind
import com.arnaudpiroelle.manga.core.utils.distinctUntilChanged
import com.arnaudpiroelle.manga.core.utils.map
import com.arnaudpiroelle.manga.core.utils.setActionBar
import com.arnaudpiroelle.manga.ui.history.HistoryContext.Action.CleanHistoryAction
import com.google.android.material.snackbar.Snackbar
import kotlinx.android.synthetic.main.fragment_listing_history.*
import kotlinx.android.synthetic.main.include_bottombar.*
import kotlinx.android.synthetic.main.include_title.*
import org.koin.androidx.viewmodel.ext.android.viewModel

class HistoryFragment : Fragment() {

    private val viewModel: HistoriesViewModel by viewModel()
    private val adapter = HistoryAdapter()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_listing_history, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        title.setText(R.string.title_history)
        setActionBar(bar)

        val linearLayoutManager = LinearLayoutManager(activity)
        list_history.layoutManager = linearLayoutManager
        list_history.adapter = adapter

        viewModel.state.map { it.visualNotification }.distinctUntilChanged().bind(this, this::onVisualNotificationChanged)

        viewModel.histories.observe(this, Observer {
            adapter.submitList(it)
        })

    }

    override fun onDestroyView() {
        list_history.adapter = null

        super.onDestroyView()
    }

    private fun checkScroll() {
        val adapter = list_history.adapter
        val layoutManager = list_history.layoutManager as LinearLayoutManager
        val totalItems = adapter?.itemCount ?: 0
        val findLastCompletelyVisibleItemPosition = layoutManager.findLastCompletelyVisibleItemPosition()

        list_history.isNestedScrollingEnabled = findLastCompletelyVisibleItemPosition != totalItems - 1
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.menu_history, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_clean_history -> {
                viewModel.handle(CleanHistoryAction)
                true
            }
            else -> false
        }
    }

    private fun onVisualNotificationChanged(visualNotification: HistoryContext.VisualNotification?) {
        if (visualNotification != null) {
            Snackbar.make(bar, visualNotification.resId, Snackbar.LENGTH_SHORT).show()
        }
    }

}
